crates/wasi-component-loader/src/cache.rs

```diff
@@ -0,0 +1,182 @@
+use std::{
+    future::Future,
+    time::{Duration, Instant, SystemTime},
+};
+
+use dashmap::DashMap;
+use futures::TryFutureExt;
+use tokio::sync::{mpsc, oneshot, Semaphore};
+use ulid::Ulid;
+use wasmtime::{
+    component::{LinkerInstance, ResourceType},
+    StoreContextMut,
+};
+
+use crate::{
+    names::{CACHE_GET_FUNCTION, CACHE_RESOURCE, CACHE_SET_FUNCTION},
+    state::WasiState,
+};
+
+type WaitListReceiver = mpsc::Receiver<oneshot::Sender<Vec<u8>>>;
+type WaitListSender = mpsc::Sender<oneshot::Sender<Vec<u8>>>;
+
+pub(crate) fn inject_mapping(types: &mut LinkerInstance<'_, WasiState>) -> crate::Result<()> {
+    types.resource(CACHE_RESOURCE, ResourceType::host::<()>(), |_, _| Ok(()))?;
+    types.func_wrap_async(CACHE_GET_FUNCTION, cache_get)?;
+    types.func_wrap_async(CACHE_SET_FUNCTION, cache_set)?;
+
+    Ok(())
+}
+
+type CacheGetResult<'a> = Box<dyn Future<Output = anyhow::Result<(Option<Vec<u8>>,)>> + Send + 'a>;
+type CacheSetResult<'a> = Box<dyn Future<Output = anyhow::Result<()>> + Send + 'a>;
+
+fn cache_get(store: StoreContextMut<'_, WasiState>, (key,): (String,)) -> CacheGetResult<'_> {
+    Box::new(async move {
+        let value = store.data().cache().get(&key).await;
+
+        Ok((value,))
+    })
+}
+
+fn cache_set(
+    store: StoreContextMut<'_, WasiState>,
+    (key, value, ttl_ms): (String, Vec<u8>, Option<u64>),
+) -> CacheSetResult<'_> {
+    Box::new(async move {
+        store.data().cache().set(&key, value, ttl_ms).await;
+        Ok(())
+    })
+}
+
+pub(crate) struct Cache {
+    cache: DashMap<String, CachedValue>,
+    semaphore: Semaphore,
+    wait_list: DashMap<String, (Ulid, WaitListSender, WaitListReceiver)>,
+}
+
+struct CachedValue {
+    data: Vec<u8>,
+    expires_at: Option<Instant>,
+}
+
+impl Cache {
+    pub fn new() -> Self {
+        Self {
+            cache: DashMap::new(),
+            semaphore: Semaphore::new(1024),
+            wait_list: DashMap::new(),
+        }
+    }
+
+    /// Gets a value from the cache by key. If this function returns None, the caller must set a new one.
```

Comment:
What happens if the caller doesn't do it? If multiple callers call get at the same time, should they all try to set?

Reply:
The first caller will create a waiting list, the second one will yield until somebody calls set. The set call is by us in the sdk, and that's the only place which can do this trick. The user only has get with an init.

If this is not done correctly, the subsequent calls yield until the timeout is hit and return None.